### ISSUE TITLE
Discrepancy between loss when called from `tf.keras.losses.BinaryCrossentropy()(true, pred)` and verbose output from model.fit() or hand-computation

### ISSUE THREAD

**Issue by cranedroesch** [tensorflow/tensorflow#43143]
I'm struggling to understand the behavior of `tf.keras.losses.BinaryCrossentropy()(true, pred)` -- I can't reproduce it's behavior from first principles.  Here's a MWE with a very simple two-output loss:

```
import tensorflow as tf
import numpy as np
from tensorflow.keras.layers import Dense
from tensorflow.keras import Model, Input

np.random.seed(8675309)
X = np.random.normal(size = 1000).reshape(100,10)
B = np.random.normal(size = 20).reshape(10,2)
Y = np.sin(X @ B ) @ np.array([1,0,0,-1]).reshape(2,2) >0

i = Input(10)
l = Dense(5, activation = 'relu')(i)
o1 = Dense(1, name = "one", activation = 'sigmoid')(l)
o2 = Dense(1, name = "two", activation = 'sigmoid')(l)
m = Model(i, [o1, o2])

loss_object = tf.keras.losses.BinaryCrossentropy(from_logits=False)

m.compile(optimizer=tf.keras.optimizers.Adam(1e-4),
              loss={'one': tf.keras.losses.BinaryCrossentropy(from_logits=False),
                    'two': tf.keras.losses.BinaryCrossentropy(from_logits=False)})

m.fit(X, [Y[:,0], Y[:,1]],
      batch_size=1000,
      epochs=3,
      verbose=1)
pred = m.predict(X)
loss = loss_object(Y, pred)
```

which gives the following output:

```
Train on 100 samples
Epoch 1/3
100/100 [==============================] - 1s 10ms/sample - loss: 1.5362 - one_loss: 0.8195 - two_loss: 0.7167
Epoch 2/3
100/100 [==============================] - 0s 47us/sample - loss: 1.5359 - one_loss: 0.8193 - two_loss: 0.7166
Epoch 3/3
100/100 [==============================] - 0s 35us/sample - loss: 1.5357 - one_loss: 0.8192 - two_loss: 0.7166


The loss from the loss object is 0.7730987071990967
```

The losses don't match.  To see what's going on, I'll compute entropy by hand:

```
P = np.concatenate([pred[0], pred[1]], axis = 1)
YY = np.concatenate([Y, 1-Y], axis = 1)
PP = np.concatenate([P, 1-P], axis = 1)

(YY*np.log(PP)).sum(axis=1).mean()
Out[47]: -1.5354833577014506

(YY[:,[0,2]]*np.log(PP[:,[0,2]])).sum(axis=1).mean()
Out[48]: -0.8189752248860895

(YY[:,[1,3]]*np.log(PP[:,[1,3]])).sum(axis=1).mean()
Out[49]: -0.7165081328153611
```

The results match the verbose output, but not the output from the loss object.  (Aside from a fairly large discrepancy in the third-ish decimal place???)

Would appreciate an explaination of this behavior.  

I have looked into the loss code on github, and it confuses me further:  

 ```
Example subclass implementation:
  ```python
  class MeanSquaredError(Loss):
    def call(self, y_true, y_pred):
      y_pred = tf.convert_to_tensor_v2(y_pred)
      y_true = tf.cast(y_true, y_pred.dtype)
      return tf.reduce_mean(math_ops.square(y_pred - y_true), axis=-1)
  ```
Converting a list to a tensor concatenates it along the first dimension, irrespective of the shape of the other input.  The rest of the github page doesn't make it clear how the entropy is actually calculated from that, or at least I can't see it.  binary cross entropy appears to be a super class, with the actual implementation somewhere else.

```

**Comment by bhack** [comment]
Generally you can't exactly match loss output on fit and loss on evaluate. See https://github.com/tensorflow/tensorflow/issues/29964#issuecomment-505498333

**Comment by cranedroesch** [comment]
> Generally you can't exactly match loss output on fit and loss on evaluate. See [#29964 (comment)](https://github.com/tensorflow/tensorflow/issues/29964#issuecomment-505498333)

That explains the discrepancy in the third-ish decimal place: the loss in the verbose output is the penultimate loss before the last gradient update.  

But that is not the main issue.  The main problem is that the loss from this:

`
loss = loss_object(Y, pred)
`

is VERY DIFFERENT from the loss that I compute by hand:

```
P = np.concatenate([pred[0], pred[1]], axis = 1)
YY = np.concatenate([Y, 1-Y], axis = 1)
PP = np.concatenate([P, 1-P], axis = 1)

(YY*np.log(PP)).sum(axis=1).mean()
Out[47]: -1.5354833577014506

(YY[:,[0,2]]*np.log(PP[:,[0,2]])).sum(axis=1).mean()
Out[48]: -0.8189752248860895

(YY[:,[1,3]]*np.log(PP[:,[1,3]])).sum(axis=1).mean()
Out[49]: -0.7165081328153611
```

Moreover, the loss that I compute by hand is similar (but not exactly the same) as the loss from the verbose output.  I guess the small dissimilarity comes from the fact that the verbose output reflects the penultimate loss, from which the gradients were computed.

**Comment by cranedroesch** [comment]
Update:  here are the results of about 50 experiments, using the loss object and using hand-computed entropy.  

(they are from my actual use-case instead of the dummy MWE in the original post)

![image](https://user-images.githubusercontent.com/13458532/93025146-937e9700-f5c9-11ea-97b4-8b19bdfba196.png)

It looks like there is a 1:1 correspondence between the loss out of the loss object, and the hand-computed loss.  It must be a scaling issue.  Someone care to shed light on it?

**Comment by bhack** [comment]
About your example here you can print the evaluation total loss, loss one and, loss two:

```
import tensorflow as tf
import numpy as np
from tensorflow.keras.layers import Dense
from tensorflow.keras import Model, Input
from tensorflow.keras.callbacks import Callback

np.random.seed(1)
tf.random.set_seed(1)

class CustomMonitoring(Callback):
    def on_train_batch_end(self, batch, logs=None):
        loss = self.model.evaluate(X,  [Y[:,0], Y[:,1]])
        print('For end batch {}, loss is {:7.4f} - one {:7.4f} - two {:7.4f}.'.format(batch, loss[0],loss[1],loss[2]))

X = np.random.normal(size = 1000).reshape(100,10)
Y = np.sin(X @ B ) @ np.array([1,0,0,-1]).reshape(2,2) >0
i = Input(10)
l = Dense(5, activation = 'relu')(i)
o1 = Dense(1, name = "one", activation = 'sigmoid')(l)
o2 = Dense(1, name = "two", activation = 'sigmoid')(l)
m = Model(i, [o1, o2])


loss_object = tf.keras.losses.BinaryCrossentropy(from_logits=False)

m.compile(optimizer=tf.keras.optimizers.Adam(1e-4),
              loss={'one': tf.keras.losses.BinaryCrossentropy(from_logits=False),
                    'two': tf.keras.losses.BinaryCrossentropy(from_logits=False)})
Y_0 = Y[:,0]
Y_1 = Y[:,1]
m.fit(X, [Y_0, Y_1],
      batch_size=100,
      epochs=3,
      verbose=1,
      shuffle=False,
      callbacks=[CustomMonitoring()])
```

You told that the main problem is `loss = loss_object(Y, pred)`. So you can reproduce `For end batch` output losses if you reproduce with your manual code all the fundamental steps in the `evaluate` pipeline (so please double check the flow).

If you see the core of the evaluation `test_step` there is relative your manual model `pred`:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L1154

As the next line is going to call `self.compiled_loss` internally you can follow the relative `loss_object` call at:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/compile_utils.py#L203
